stake-pool/js/src/utils/stake.ts

```diff
@@ -71,55 +94,60 @@ export async function prepareWithdrawAccounts(
       });
     }
 
-    const transientStakeAccountAddress = await findTransientStakeProgramAddress(
-      STAKE_POOL_PROGRAM_ID,
-      validator.voteAccountAddress,
-      stakePoolAddress,
-      validator.transientSeedSuffixStart,
-    );
-
-    if (!validator.transientStakeLamports?.isZero()) {
+    const transientStakeLamports = validator.transientStakeLamports.toNumber() - minBalance;
+    if (transientStakeLamports > 0) {
+      const transientStakeAccountAddress = await findTransientStakeProgramAddress(
+        STAKE_POOL_PROGRAM_ID,
+        validator.voteAccountAddress,
+        stakePoolAddress,
+        validator.transientSeedSuffixStart,
+      );
       accounts.push({
         type: 'transient',
         voteAddress: validator.voteAccountAddress,
         stakeAddress: transientStakeAccountAddress,
-        lamports: validator.transientStakeLamports.toNumber(),
+        lamports: transientStakeLamports,
       });
     }
   }
 
   // Sort from highest to lowest balance
-  accounts = accounts.sort((a, b) => b.lamports - a.lamports);
+  accounts = accounts.sort(compareFn ? compareFn : (a, b) => b.lamports - a.lamports);
 
   const reserveStake = await connection.getAccountInfo(stakePool.reserveStake);
-  if (reserveStake && reserveStake.lamports > 0) {
-    console.log('Reserve Stake: ', reserveStake.lamports);
+  const reserveStakeBalance = (reserveStake?.lamports ?? 0) - minBalanceForRentExemption - 1;
+  if (reserveStakeBalance > 0) {
     accounts.push({
       type: 'reserve',
       stakeAddress: stakePool.reserveStake,
-      lamports: reserveStake?.lamports,
+      lamports: reserveStakeBalance,
     });
   }
 
   // Prepare the list of accounts to withdraw from
   const withdrawFrom: WithdrawAccount[] = [];
   let remainingAmount = amount;
 
+  const fee = stakePool.stakeWithdrawalFee;
+  const inverseFee: Fee = {
+    numerator: fee.denominator.sub(fee.numerator),
+    denominator: fee.denominator,
+  };
+
   for (const type of ['preferred', 'active', 'transient', 'reserve']) {
-    const filteredAccounts = accounts.filter(a => a.type == type);
+    const filteredAccounts = accounts.filter((a) => a.type == type);
 
-    for (const {stakeAddress, voteAddress, lamports} of filteredAccounts) {
-      let availableForWithdrawal = Math.floor(
-        calcPoolTokensForDeposit(stakePool, lamports),
-      );
-      if (!stakePool.stakeWithdrawalFee.denominator.isZero()) {
+    for (const { stakeAddress, voteAddress, lamports } of filteredAccounts) {
+      if (lamports <= minBalance) {
```

Comment:
I believe this condition should be changed to `(lamports <= minBalance && type == "transient")` because I think @joncinque was saying that the minimum balance part only applies to transient stake accounts.

Reply:
With this fix, everything seems to be working on devnet.

Reply:
The reserve part is still problematic, but from inspecting the issue, it seems like there's not much that can be done in the bindings themselves (seems to be an issue in the program itself). For instance, with completely empty validator stake accounts and all the transient stake set to the minimum balance, the program still returns:

```
Program log: Error withdrawing from reserve: validator stake accounts have lamports available, please use those first.
Program log: Error: The lamports in the validator stake account is not equal to the minimum
```

It seems like it might actually be impossible to withdraw stake from the reserve account even when all other accounts have been fully depleted! In that case, it's probably not worth spending too much time trying to fix that on the bindings side.

Reply:
I can confirm the reserve part does have a corner case, since it's impossible to totally drain a transient account.  I think it's possible to relax that requirement, but I have to think through the implications.

Reply:
'use_reserve' functionality is just reimplemented bases on cli version

https://github.com/solana-labs/solana-program-library/blob/3d92f8f4dc4fab63b9eb68f33185d9c2829abcfe/stake-pool/cli/src/main.rs#L1373

@joncinque  should I also remove it from cli or make it still sense there?